File: Src/Engine/Editor/GUI/Docking/DockPanel.cpp

#include "DockPanel.h"

#include "DockPanelProxy.h"
#include "DockWindow.h"
#include "Runtime/UI/GUI/RootControl.h"
#include "Runtime/UI/GUI/Panels/SplitPanel.h"

namespace SE::Editor
{
	DockPanel::DockPanel(DockPanel* parentPanel)
	{
		AutoFocus = false;

		_parentPanel = parentPanel;
		if (_parentPanel != nullptr)
		{
			_parentPanel->_childPanels.Add(this);
		}

		AnchorPreset = AnchorPresets::StretchAll;
		Offsets = Margin::Zero;
	}

	bool DockPanel::CloseAll(ClosingReason reason)
	{
		while (_tabs.Count() > 0)
		{
			if (!_tabs[0]->Close(reason))
				return false;
		}

		return true;
	}

	DockWindow* DockPanel::GetTab(int tabIndex)
	{
		return _tabs[tabIndex];
	}

	int DockPanel::GetTabIndex(Editor::DockWindow* tab)
	{
		return _tabs.Find(tab);
	}

	bool DockPanel::ContainsTab(Editor::DockWindow* tab)
	{
		return _tabs.Contains(tab);
	}

	void DockPanel::SelectTab(int tabIndex)
	{
		::SE::Editor::DockWindow* tab = nullptr;
		if (tabIndex >= 0 && _tabs.Count() > tabIndex && _tabs.Count() > 0)
			tab = _tabs[tabIndex];
		SelectTab(tab);
	}

	void DockPanel::SelectTab(Editor::DockWindow* tab, bool autoFocus)
	{
		// Check if tab will change
		if (_selectedTab != tab)
		{
			// Change
			ContainerControl* proxy;
			if (_selectedTab != nullptr)
			{
				proxy = _selectedTab->Parent;
				_selectedTab->Parent = nullptr;
			}
			else
			{
				CreateTabsProxy();
				proxy = _tabsProxy;
			}
			_selectedTab = tab;
			if (_selectedTab != nullptr)
			{
				_selectedTab->UnlockChildrenRecursive();
				_selectedTab->Parent = proxy;
				if (autoFocus)
					_selectedTab->Focus();
			}
			OnSelectedTabChanged();
		}
		else if (autoFocus && _selectedTab != nullptr && !_selectedTab->ContainsFocus())
		{
			_selectedTab->Focus();
		}
	}

	DockPanel* DockPanel::HitTest(Float2& position)
	{
		// Get parent window and transform point position into local coordinates system
		RootControl* parentWin = Root;
		if (parentWin == nullptr)
			return nullptr;
		auto clientPos = parentWin->PointFromScreen(position);
		auto localPos = PointFromWindow(clientPos);

		// Early out
		if (localPos.x < 0 || localPos.y < 0)
			return nullptr;
		if (localPos.x > Width || localPos.y > Height)
			return nullptr;

		// Test all docked controls (find the smallest one)
		float sizeLengthSquared = Max_float;
		DockPanel* result = nullptr;
		for (int i = 0; i < _childPanels.Count(); i++)
		{
			auto panel = _childPanels[i]->HitTest(position);
			if (panel != nullptr)
			{
				auto sizeLen = panel->Size.operator->().LengthSquared();
				if (sizeLen < sizeLengthSquared)
				{
					sizeLengthSquared = sizeLen;
					result = panel;
				}
			}
		}
		if (result != nullptr)
			return result;

		// Itself
		return this;
	}

	DockState DockPanel::TryGetDockState(float& splitterValue)
	{
		DockState result = DockState::Unknown;
		splitterValue = DefaultSplitterValue;

		if (HasParent())
		{
			SplitPanel* splitter;
			if (TypeTryCast<SplitPanel>(Parent->Parent, splitter))
			{
				splitterValue = splitter->SplitterValue;
				if (Parent == splitter->Panel1)
				{
					if (splitter->Orientation == Orientation::Horizontal)
						result = DockState::DockLeft;
					else
						result = DockState::DockTop;
				}
				else
				{
					if (splitter->Orientation == Orientation::Horizontal)
						result = DockState::DockRight;
					else
						result = DockState::DockBottom;
					splitterValue = 1 - splitterValue;
				}
			}
		}

		return result;
	}

	DockPanel* DockPanel::CreateChildPanel(DockState state, float splitterValue)
	{
		CreateTabsProxy();

		// Create child dock panel
		DockPanel* dockPanel = New<DockPanel>(this);

		// Switch dock mode
		Control* c1 = nullptr;
		Control* c2 = nullptr;
		Orientation o;
		switch (state)
		{
		case DockState::DockTop:
		{
			o = Orientation::Vertical;
			c1 = dockPanel;
			c2 = _tabsProxy;
			break;
		}
		case DockState::DockBottom:
		{
			splitterValue = 1 - splitterValue;
			o = Orientation::Vertical;
			c1 = _tabsProxy;
			c2 = dockPanel;
			break;
		}
		case DockState::DockLeft:
		{
			o = Orientation::Horizontal;
			c1 = dockPanel;
			c2 = _tabsProxy;
			break;
		}
		case DockState::DockRight:
		{
			splitterValue = 1 - splitterValue;
			o = Orientation::Horizontal;
			c1 = _tabsProxy;
			c2 = dockPanel;
			break;
		}
		}

		// Create splitter and link controls
		ContainerControl* parent = _tabsProxy->Parent;
		SplitPanel* splitter = New<SplitPanel>(o, ScrollBars::None, ScrollBars::None);
		splitter->AnchorPreset = AnchorPresets::StretchAll;
		splitter->Offsets = Margin::Zero;
		splitter->SplitterValue = splitterValue;

		splitter->Panel1->AddChild(c1);
		splitter->Panel2->AddChild(c2);
		parent->AddChild(splitter);

		// Update
		splitter->UnlockChildrenRecursive();
		splitter->PerformLayout();

		return dockPanel;
	}

	void DockPanel::MoveTabLeft(int index)
	{
		if (index > 0)
		{
			auto tab = _tabs[index];
			_tabs.RemoveAt(index);
			_tabs.Insert(index - 1, tab);
		}
	}
	void DockPanel::MoveTabRight(int index)
	{
		if (index < _tabs.Count() - 2)
		{
			auto tab = _tabs[index];
			_tabs.RemoveAt(index);
			_tabs.Insert(index + 1, tab);
		}
	}

	void DockPanel::OnDestroy()
	{
		if (_parentPanel != nullptr)
		{
			_parentPanel->_childPanels.Remove(this);
		}

		ContainerControl::OnDestroy();

		// Clear other tabs (not in the view)
		for (int i = 0; i < _tabs.Count(); i++)
		{
			if (!_tabs[i]->IsDisposing)
			{
				_tabs[i]->Dispose();
			}
		}
	}

	void DockPanel::OnLastTabRemoved()
	{
		// Check if dock panel is linked to the split panel
		if (HasParent())
		{
			SplitPanel* splitter;
			if (TypeTryCast<SplitPanel>(Parent->Parent, splitter))
			{
				// Check if there is another nested dock panel inside this dock panel and extract it here
				auto childPanels = _childPanels;
				if (childPanels.Count() != 0)
				{
					// Move tabs from child panels into this one
					Editor::DockWindow* selectedTab = nullptr;
					for (DockPanel* childPanel : childPanels)
					{
						List<Editor::DockWindow*>& childPanelTabs = childPanel->Tabs;
						for (int i = 0; i < childPanelTabs.Count(); i++)
						{
							Editor::DockWindow* childPanelTab = childPanelTabs[i];
							if (selectedTab == nullptr && childPanelTab->IsSelected)
								selectedTab = childPanelTab;
							childPanel->UndockWindow(childPanelTab);
							AddTab(childPanelTab, false);
						}
					}
					if (selectedTab != nullptr)
						SelectTab(selectedTab);
				}
				else
				{
					// Unlink splitter
					ContainerControl* splitterParent = splitter->Parent;
					ENGINE_ASSERT(splitterParent != nullptr);
					splitter->Parent = nullptr;

					// Move controls from second split panel to the split panel parent
					Panel* scrPanel = Parent == splitter->Panel2 ? splitter->Panel1 : splitter->Panel2;
					int srcPanelChildrenCount = scrPanel->ChildrenCount();
					for (int i = srcPanelChildrenCount - 1; i >= 0 && scrPanel->ChildrenCount() > 0; i--)
					{
						scrPanel->GetChild(i)->Parent = splitterParent;
					}

					ENGINE_ASSERT(scrPanel->ChildrenCount() == 0);
					ENGINE_ASSERT(splitterParent->ChildrenCount() == srcPanelChildrenCount);

					// Delete
					splitter->Dispose();
					Delete(splitter);
				}
			}
			else if (!IsMaster)
			{
				ENGINE_UNREACHABLE_CODE()
			}
		}
		else if (!IsMaster)
		{
			ENGINE_UNREACHABLE_CODE()
		}
	}

	void DockPanel::DockWindow(DockState state, Editor::DockWindow* window, bool autoSelect, float splitterValue)
	{
		CreateTabsProxy();

		// Check if dock like a tab or not
		if (state == DockState::DockFill)
		{
			// Add tab
			AddTab(window, autoSelect);
		}
		else
		{
			// Create child panel
			auto dockPanel = CreateChildPanel(state, splitterValue != 0.0 ? splitterValue : DefaultSplitterValue);

			// Dock window as a tab in a child panel
			dockPanel->DockWindow(DockState::DockFill, window);
		}
	}

	void DockPanel::UndockWindow(::SE::Editor::DockWindow* window)
	{
		int index = GetTabIndex(window);
		ENGINE_ASSERT(index >= 0);

		// Check if tab was selected
		if (window == SelectedTab)
		{
			// Change selection
			if (index == 0 && TabsCount > 1)
				SelectTab(1);
			else
				SelectTab(index - 1);
		}

		// Undock
		_tabs.RemoveAt(index);
		window->ParentDockPanel = nullptr;

		// Check if has no more tabs
		if (_tabs.Count() == 0)
		{
			OnLastTabRemoved();
		}
		else
		{
			// Update
			PerformLayout();
		}
	}

	void DockPanel::AddTab(Editor::DockWindow* window, bool autoSelect)
	{
		_tabs.Add(window);
		window->ParentDockPanel = this;
		if (autoSelect)
			SelectTab(window);
	}

	void DockPanel::CreateTabsProxy()
	{
		if (_tabsProxy == nullptr)
		{
			// Create proxy and make set simple full dock
			_tabsProxy = New<DockPanelProxy>(this);
			_tabsProxy->Parent = this;

			_tabsProxy->UnlockChildrenRecursive();
		}
	}

	int DockPanel::__GetSelectedTabIndex()
	{
		return _tabs.Find(_selectedTab);
	}

	Rectangle DockPanel::__GetDockAreaBounds()
	{
		RootControl* parentWin = Root;
		ENGINE_ASSERT_M(parentWin != nullptr, SE_TEXT("Missing parent window."));

		Control* control = _tabsProxy != nullptr ? static_cast<Control*>(_tabsProxy) : this;
		Float2 clientPos = control->PointToWindow(Float2::Zero);
		return Rectangle(parentWin->PointToScreen(clientPos), Float2(GetDpiScale()) * control->Size);
	}
} // SE

File: Src/Engine/Editor/GUI/Docking/DockPanel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using SE.GUI;

namespace SE.Editor.GUI
{
    public class DockPanel : ContainerControl
    {
        public const float DefaultHeaderHeight = 25.0f;
        public const float DefaultTextMargin = 2.0f;
        public const float DefaultButtonsSize = 12.0f;
        public const float DefaultButtonsMargin = 4.0f;
        public const float DefaultSplitterValue = 0.25f;

        private readonly List<DockPanel> m_ChildPanels = new List<DockPanel>();
        private readonly List<DockWindow> m_Tabs = new List<DockWindow>();
        private DockWindow? m_SelectedTab;
        private DockPanelProxy? m_TabsProxy;

        public DockPanel(DockPanel? parentPanel = null)
            : base(new Rectangle(0, 0, 300, 200))
        {
            AutoFocus = false;
            AnchorMin = Float2.Zero;
            AnchorMax = Float2.One;
            Offsets = Margin.Zero;
            ParentDockPanel = parentPanel;
            parentPanel?.m_ChildPanels.Add(this);
        }

        public virtual bool IsMaster => false;
        public virtual bool IsFloating => false;
        public Rectangle DockAreaBounds
        {
            get
            {
                Control control = m_TabsProxy ?? this;
                Rectangle bounds = control.ScreenBounds;
                if (Root is WindowRootControl root)
                {
                    return new Rectangle(
                        root.Window.ClientToScreen(bounds.Location * root.DpiScale),
                        bounds.Size * root.DpiScale);
                }
                return bounds;
            }
        }
        public IReadOnlyList<DockPanel> ChildPanels => m_ChildPanels;
        public int ChildPanelsCount => m_ChildPanels.Count;
        public IReadOnlyList<DockWindow> Tabs => m_Tabs;
        public int TabsCount => m_Tabs.Count;
        public DockWindow? SelectedTab => m_SelectedTab;
        public DockWindow? FirstTab => m_Tabs.Count > 0 ? m_Tabs[0] : null;
        public DockWindow? LastTab => m_Tabs.Count > 0 ? m_Tabs[m_Tabs.Count - 1] : null;
        public DockPanel? ParentDockPanel { get; }
        public DockPanelProxy? TabsProxy => m_TabsProxy;

        public int SelectedTabIndex
        {
            get => m_SelectedTab != null ? m_Tabs.IndexOf(m_SelectedTab) : -1;
            set => SelectTab(value);
        }

        public bool CloseAll(ClosingReason reason = ClosingReason.CloseEvent)
        {
            // Deliberately follows the native active control flow and return-value contract.
            while (m_Tabs.Count > 0)
            {
                if (!m_Tabs[0].Close(reason))
                    return false;
            }
            return true;
        }

        public DockWindow GetTab(int tabIndex) => m_Tabs[tabIndex];
        public int GetTabIndex(DockWindow tab) => m_Tabs.IndexOf(tab);
        public bool ContainsTab(DockWindow tab) => m_Tabs.Contains(tab);

        public void SelectTab(int tabIndex)
        {
            DockWindow? tab = tabIndex >= 0 && tabIndex < m_Tabs.Count ? m_Tabs[tabIndex] : null;
            SelectTab(tab);
        }

        public void SelectTab(DockWindow? tab, bool autoFocus = true)
        {
            if (!ReferenceEquals(m_SelectedTab, tab))
            {
                ContainerControl proxy;
                if (m_SelectedTab != null)
                {
                    proxy = m_SelectedTab.Parent!;
                    proxy.RemoveChild(m_SelectedTab);
                }
                else
                {
                    proxy = CreateTabsProxy();
                }

                m_SelectedTab = tab;
                if (m_SelectedTab != null)
                {
                    m_SelectedTab.UnlockChildrenRecursive();
                    proxy.AddChild(m_SelectedTab);
                    if (autoFocus)
                        m_SelectedTab.Focus();
                }
                OnSelectedTabChanged();
            }
            else if (autoFocus && m_SelectedTab != null && !m_SelectedTab.ContainsFocus)
            {
                m_SelectedTab.Focus();
            }
        }

        public override DockPanel? HitTest(Float2 position)
        {
            Float2 rootPosition = position;
            if (Root is WindowRootControl root)
                rootPosition = root.Window.ScreenToClient(position) / Math.Max(root.DpiScale, 0.001f);
            if (!ScreenBounds.Contains(rootPosition))
                return null;

            DockPanel? result = null;
            float smallestSize = float.MaxValue;
            foreach (DockPanel child in m_ChildPanels)
            {
                DockPanel? hit = child.HitTest(position);
                if (hit == null)
                    continue;
                float size = hit.Width * hit.Width + hit.Height * hit.Height;
                if (size < smallestSize)
                {
                    smallestSize = size;
                    result = hit;
                }
            }
            return result ?? this;
        }

        public virtual DockState TryGetDockState(out float splitterValue)
        {
            splitterValue = DefaultSplitterValue;
            if (Parent is Panel host && host.Parent is SplitPanel splitter)
            {
                splitterValue = splitter.SplitterValue;
                if (ReferenceEquals(host, splitter.Panel1))
                    return splitter.Orientation == Orientation.Horizontal ? DockState.DockLeft : DockState.DockTop;

                splitterValue = 1.0f - splitterValue;
                return splitter.Orientation == Orientation.Horizontal ? DockState.DockRight : DockState.DockBottom;
            }
            return DockState.Unknown;
        }

        public DockPanel CreateChildPanel(DockState state, float splitterValue)
        {
            DockPanelProxy tabsProxy = CreateTabsProxy();
            ContainerControl parent = tabsProxy.Parent ?? this;
            DockPanel child = new DockPanel(this);
            Control first;
            Control second;
            Orientation orientation;

            switch (state)
            {
            case DockState.DockTop:
                orientation = Orientation.Vertical;
                first = child;
                second = tabsProxy;
                break;
            case DockState.DockBottom:
                splitterValue = 1.0f - splitterValue;
                orientation = Orientation.Vertical;
                first = tabsProxy;
                second = child;
                break;
            case DockState.DockLeft:
                orientation = Orientation.Horizontal;
                first = child;
                second = tabsProxy;
                break;
            case DockState.DockRight:
                splitterValue = 1.0f - splitterValue;
                orientation = Orientation.Horizontal;
                first = tabsProxy;
                second = child;
                break;
            default:
                throw new ArgumentOutOfRangeException(nameof(state));
            }

            SplitPanel splitter = new SplitPanel(orientation) { SplitterValue = splitterValue };
            splitter.AnchorMin = Float2.Zero;
            splitter.AnchorMax = Float2.One;
            splitter.Offsets = Margin.Zero;
            splitter.Panel1.AddChild(first);
            splitter.Panel2.AddChild(second);
            parent.AddChild(splitter);
            splitter.UnlockChildrenRecursive();
            splitter.PerformLayout();
            return child;
        }

        public virtual void DockWindowInternal(DockState state, DockWindow window, bool autoSelect = true, float splitterValue = 0)
            => DockWindow(state, window, autoSelect, splitterValue);
        public void RemoveIt() => OnLastTabRemoved();
        public void UndockWindowInternal(DockWindow window) => UndockWindow(window);

        public void MoveTabLeft(int index)
        {
            if (index <= 0 || index >= m_Tabs.Count)
                return;
            DockWindow tab = m_Tabs[index];
            m_Tabs.RemoveAt(index);
            m_Tabs.Insert(index - 1, tab);
        }

        public void MoveTabRight(int index)
        {
            // Native currently stops one slot before the final tab; preserve that active behavior.
            if (index < 0 || index >= m_Tabs.Count - 2)
                return;
            DockWindow tab = m_Tabs[index];
            m_Tabs.RemoveAt(index);
            m_Tabs.Insert(index + 1, tab);
        }

        protected virtual void OnLastTabRemoved()
        {
            if (Parent is not Panel host || host.Parent is not SplitPanel splitter)
                return;

            if (m_ChildPanels.Count > 0)
            {
                DockWindow? selected = null;
                foreach (DockPanel child in m_ChildPanels.ToArray())
                {
                    // Native code walks the live list forward while UndockWindow mutates it.
                    for (int index = 0; index < child.m_Tabs.Count; index++)
                    {
                        DockWindow tab = child.m_Tabs[index];
                        if (selected == null && tab.IsSelected)
                            selected = tab;
                        child.UndockWindow(tab);
                        AddTab(tab, false);
                    }
                }
                if (selected != null)
                    SelectTab(selected);
                return;
            }

            ContainerControl? splitterParent = splitter.Parent;
            if (splitterParent == null)
                return;
            Panel source = ReferenceEquals(host, splitter.Panel2) ? splitter.Panel1 : splitter.Panel2;
            Control[] remaining = source.Children
                .Where(control => !ReferenceEquals(control, source.VerticalScrollBar) &&
                                  !ReferenceEquals(control, source.HorizontalScrollBar))
                .ToArray();
            splitterParent.RemoveChild(splitter);
            for (int index = remaining.Length - 1; index >= 0; index--)
                splitterParent.AddChild(remaining[index]);
            splitter.Dispose();
        }

        protected virtual void DockWindow(DockState state, DockWindow window, bool autoSelect = true, float splitterValue = 0)
        {
            CreateTabsProxy();
            if (state == DockState.DockFill)
            {
                AddTab(window, autoSelect);
            }
            else
            {
                DockPanel child = CreateChildPanel(state, splitterValue != 0.0f ? splitterValue : DefaultSplitterValue);
                child.DockWindow(DockState.DockFill, window);
            }
        }

        protected virtual void UndockWindow(DockWindow window)
        {
            int index = GetTabIndex(window);
            if (index < 0)
                return;

            if (ReferenceEquals(window, m_SelectedTab))
                SelectTab(index == 0 && m_Tabs.Count > 1 ? 1 : index - 1);

            m_Tabs.RemoveAt(index);
            window.ParentDockPanel = null;
            if (m_Tabs.Count == 0)
                OnLastTabRemoved();
            else
                PerformLayout();
        }

        protected virtual void AddTab(DockWindow window, bool autoSelect = true)
        {
            m_Tabs.Add(window);
            window.ParentDockPanel = this;
            if (autoSelect)
                SelectTab(window);
        }

        protected virtual void OnSelectedTabChanged()
        {
        }

        protected override void OnLayoutChildren()
        {
            m_TabsProxy?.SetBounds(0.0f, 0.0f, Width, Height);
        }

        protected override void OnDispose()
        {
            if (ParentDockPanel != null)
                ParentDockPanel.m_ChildPanels.Remove(this);

            base.OnDispose();

            // Native DockPanel destroys tabs that were not attached to the active proxy view.
            for (int index = 0; index < m_Tabs.Count; index++)
            {
                DockWindow tab = m_Tabs[index];
                tab.ParentDockPanel = null;
                if (!tab.IsDisposed)
                    tab.Dispose();
            }
            m_Tabs.Clear();
            m_ChildPanels.Clear();
            m_SelectedTab = null;
            m_TabsProxy = null;
        }

        private DockPanelProxy CreateTabsProxy()
        {
            if (m_TabsProxy == null)
            {
                m_TabsProxy = new DockPanelProxy(this);
                AddChild(m_TabsProxy);
                m_TabsProxy.UnlockChildrenRecursive();
            }
            return m_TabsProxy;
        }
    }
}
